### PR TITLE
Improve handling of Windows Metadata sources, add error diagnostic

### LINKED ISSUE #481: cswinrt.exe errors should return non-zero exit code, fail the build

### PR DESCRIPTION
**Pull request by Scottj1s**
fixes #481, #507 

also validates support of contracts package in the sample

### CODE REVIEW

File: nuget/Microsoft.Windows.CsWinRT.targets [hunk 1]
```diff
@@ -1,101 +1,108 @@
...
-    <CompilerVisibleProperty Include="CsWinRTComponent" />
-    <CompilerVisibleProperty Include="CsWinRTEnableLogging" />
-    <CompilerVisibleProperty Include="GeneratedFilesDir" />
-    <CompilerVisibleProperty Include="CsWinRTExe" />
-  </ItemGroup>
-
-  <Import Project="$(MSBuildThisFileDirectory)Microsoft.Windows.CsWinRT.Prerelease.targets" Condition="Exists('$(MSBuildThisFileDirectory)Microsoft.Windows.CsWinRT.Prerelease.targets')"/>
-
-</Project>
+<!--
```
Comment (Scottj1s) [inline nuget/Microsoft.Windows.CsWinRT.targets:1]: sorry about the whitespace noise

File: nuget/Microsoft.Windows.CsWinRT.targets [hunk 2]
```diff
@@ -1,101 +1,108 @@
...
+      <CsWinRTCommand>$(CsWinRTExe) %40"$(CsWinRTResponseFile)"</CsWinRTCommand>
+      <CsWinRTWindowsMetadata Condition="'$(CsWinRTWindowsMetadata)' == ''">$(WindowsSDKVersion.TrimEnd('\'))</CsWinRTWindowsMetadata>
+      <CsWinRTWindowsMetadataInput Condition="'$(CsWinRTWindowsMetadata)' != ''">-input $(CsWinRTWindowsMetadata)</CsWinRTWindowsMetadataInput>
+    </PropertyGroup>
+    <ItemGroup Condition="'$(CsWinRTParams)$(CsWinRTWindowsMetadata)' == ''">
+      <CsWinRTDetectWindowsMetadata Include="@(CsWinRTInputs)" Condition="'%(CsWinRTInputs.Filename)%(CsWinRTInputs.Extension)' == 'Windows.Foundation.FoundationContract.winmd'"></CsWinRTDetectWindowsMetadata>
+      <CsWinRTDetectWindowsMetadata Include="@(CsWinRTInputs)" Condition="'%(CsWinRTInputs.Filename)%(CsWinRTInputs.Extension)' == 'Windows.winmd'"></CsWinRTDetectWindowsMetadata>
+    </ItemGroup>
+    <Error Condition="'$(CsWinRTParams)$(CsWinRTWindowsMetadata)@(CsWinRTDetectWindowsMetadata)' == ''"
+      Text="Windows Metadata not provided or detected.  See https://github.com/microsoft/CsWinRT/tree/master/nuget/readme.md" />
```
Comment (manodasanW) [inline nuget/Microsoft.Windows.CsWinRT.targets:63]: I  am concerned this error would actually break building the Windows SDK projection itself as I currently feed those WinMDs like     <CsWinRTInputs Include="$(ProjectionFolder)/FilteredSDK/**/*.winmd" />.  With this model, how should I pass them?

File: nuget/Microsoft.Windows.CsWinRT.targets [hunk 3]
```diff
@@ -1,101 +1,108 @@
...
+
+  <Target Name="CsWinRTGenerateProjection" DependsOnTargets="CsWinRTPrepareProjection;CsWinRTRemoveWinMDReferences" Condition="'$(CsWinRTGenerateProjection)' == 'true'">
+    <PropertyGroup>
+      <CsWinRTResponseFile>$(CsWinRTGeneratedFilesDir)cswinrt.rsp</CsWinRTResponseFile>
+      <CsWinRTCommand>$(CsWinRTExe) %40"$(CsWinRTResponseFile)"</CsWinRTCommand>
+      <CsWinRTWindowsMetadata Condition="'$(CsWinRTWindowsMetadata)' == ''">$(WindowsSDKVersion.TrimEnd('\'))</CsWinRTWindowsMetadata>
+      <CsWinRTWindowsMetadataInput Condition="'$(CsWinRTWindowsMetadata)' != ''">-input $(CsWinRTWindowsMetadata)</CsWinRTWindowsMetadataInput>
+    </PropertyGroup>
+    <ItemGroup Condition="'$(CsWinRTParams)$(CsWinRTWindowsMetadata)' == ''">
+      <CsWinRTDetectWindowsMetadata Include="@(CsWinRTInputs)" Condition="'%(CsWinRTInputs.Filename)%(CsWinRTInputs.Extension)' == 'Windows.Foundation.FoundationContract.winmd'"></CsWinRTDetectWindowsMetadata>
```
Comment (manodasanW) [inline nuget/Microsoft.Windows.CsWinRT.targets:59]: Where is this used other than for error?

Comment (Scottj1s) [reply nuget/Microsoft.Windows.CsWinRT.targets:59]: only for error

File: nuget/Microsoft.Windows.CsWinRT.targets (hunk 2 above)
Comment (Scottj1s) [reply nuget/Microsoft.Windows.CsWinRT.targets:63]: You're not supplying Windows.Foundation.FoundationContract.winmd or Windows.winmd as CsWinRTInputs?  if you are, it shouldn't error

Comment (manodasanW) [reply nuget/Microsoft.Windows.CsWinRT.targets:63]: I misunderstood yesterday what CsWinRTDetectWindowsMetadata was doing.  Now it seems to make sense once you mentioned that we are only using it to detect whether there is no windows winmd and then use that to error.

**Review by manodasanW — APPROVED** [review]

File: nuget/Microsoft.Windows.CsWinRT.targets (hunk 2 above)
Comment (Scottj1s) [reply nuget/Microsoft.Windows.CsWinRT.targets:63]: actually, I'm glad you brought it up - there was a problem with the conditions

**Pull request merged**